rfcs/text/0000-dss-deletion-daemon.md

```diff
@@ -0,0 +1,631 @@
+   mentioned in the requests response. For **physical deletion** requests, tombstones are placed for all files and 
+   bundles mentioned in the request response. The tombstone's content is
+   as follows:
+   ```JSON
+   { "reason": "<reason>",
+     "details": "<additional info>",
+     "email": "<requester's email>"
+   }
+   ```
+   
+   A versioned bundle tombstone will cause `GET /bundle/{uuid}` to return a 410 for that bundle version. If the tombstoned
+   bundle is the latest version then a request for `GET /bundle/{uuid}` without a version will
+   also returns a 410. If bundle tombstone does not specify a bundle version then all request for `GET /bundle/{uuid}` will
+   return 410 even if version is specified in the request. The same applies to `HEAD /bundle/{uuid}` requests, both versioned and
+   unversioned. The same applied for `GET /files/{uuid}` and `HEAD /files/{uuid}`. Bundles that have an 
+   associated tombstone are inaccessible through the API and are considered **logically deleted**. Files with an 
+   associated tombstone or not logically delete. User with a direct URL to the file can still access the data.
+   
+   A notification is sent out to subscribers of datastore deletions when a new bundle tombstone is created, so they can 
+   update their index appropriately. If the the bundle was logically deleted and later physically deleted a second
+   identical notification will be sent to subscribers. The subscribers must be able to handle these notifications 
+   idempotently.
+   
+   A **physical deletion** request of a bundles results in the bundle associated files being added to 
+   the deletion queue for regularly schedule **Physical Deletion**. A **physical deletion** request of a file result in 
+   that file being added to the to deletion queue
+
+5) On a regular schedule the deletion queue is processed by code running within a secure boundary. An administrator 
+   may manually invoke the deletion process or wait for it run at it's scheduled time once daily.
+   The deletion daemon is responsible for performing **physical deletes** of data from the deletion queue. 
+   A physical delete makes all files associated with a bundle or bundle version inaccessible. This will cause direct
+   urls to a deleted file to fail.
+   
+   A log entry is produced when ever the deletion daemon is run. It contains information on what bundles and files
+   have been logically deleted, the reason, the administrator who executed the deletion, and info about the deletion markers. 
+   The same information is stored in the **Deletion Table**. The deletion table tracks all bundles and files currently
+   pending permanent deletion. The Deletion Table contains enough information to reverse Physical Deletes before 
+   the data is Permanently deleted.
+   
+   A daily digest of bundles and files that are scheduled for deletion are sent to DCP administrator 24 hours prior to the data
```

Comment:
There are 3 stage of deletion in total. Only two stages of deletion that Users care about. Logical deletion basically the creation of a tombstone and making it invisible to API. Physical deletion is using the bucket versioning to delete the item from the bucket without completely removing it. Physical and Logical are the only ones uses should worry about. Permanent deletion uses the bucket life cycle policies to clean up deleted files once they are so old. The time between physical deletion and permanent deletion is the grace period. We know when the physical deletion occurred and how long the grace period is we can determine how long before the physically deleted data. When <24 hours remains before the grace period is up we'd like to send a notification to warn about pending permanent deletions.

<36 hours might actually be a better time window for notifications.

Reply:
I think it would get spammy if we send a digest every time a deletion occurred when it should be impossible to permanently delete something earlier then the grace period.

Reply:
Ah, that clears it up.  Yeah, that makes sense, ignore my comment then.